### PR TITLE
Adds static library support to build

### PR DESCRIPTION
**Pull request by jstebbins**
Also de-duplicates multiply defined functions.

See https://github.com/OpenVisualCloud/SVT-AV1/issues/39

### CODE REVIEW

**Comment by hassount** [comment]
@kylophone @EwoutH 
Do you think we should update travis to build both static and shared ?

**Comment by jstebbins** [comment]
By default, this PR would build both.  I have only provided a way to disable shared (-DENABLE_SHARED=OFF).  Did you also want a way to disable static?

**Comment by hassount** [comment]
@jstebbins 
Thank you very much for the PR, this looks good to me, a couple of items to change though: 
- We don't need the string functions in the library and we would like to restrict their usage to the application only. Could you please not have them as APIs and make EbString.h /c an application only file. 
- The string safe functions we got from the safe string open source library, so the license header should reflect so and can be found in the EbAppFifo.c file
- The only function that would be needed in the library is the memset function which is a safe function to use so you can change that in the library without having to carry it as an API. 

Please let me know if you need help with these changes and we could help addressing some of them.

**Comment by jstebbins** [comment]
@hassount sounds good.  It might be a few days before I can get back to it, but I'll update the PR with those changes.

**Comment by jstebbins** [comment]
This commit a6c1f81989c6cab0f477adfa867d5ff3dad2725c just introduced use of the string functions into the library (it uses strnlen_ss).  How would you like to resolve this?

**Comment by jstebbins** [comment]
For now, to resolve the conflict, I just changed EbEncHandle.c to use regular strnlen, see 1173b38. Please let me know if this is not a suitable way to handle this.

Force pushed to clean up history and resolve conflicts with master.

**Comment by hassount** [comment]
@jstebbins 
No Problem, I would also wait for #135 to avoid conflicts.

**Comment by jstebbins** [comment]
Sure, I'll just rebase this on whatever comes along.  You can merge whenever you think appropriate.

**Comment by tmatth** [comment]
> Sure, I'll just rebase this on whatever comes along. You can merge whenever you think appropriate.

Looks like #135 was merged.

**Comment by jstebbins** [comment]
Thanks for the ping.  Conflicts resolved.

I'm not sure if my changes to the time functions fits properly into the API rework that just happened.  Please let me know if that needs to be done differently.

**Comment by 1480c1** [comment]
@jstebbins can you confirm if the conflicts are resolved, it will still build static libs? Right now, if I build using mingw-w64, it will produce .dll and .dll.a which I personally feel is undesirable.

**Comment by jstebbins** [comment]
The latest set of conflicts are not resolved as github indicates.  I'll fix that when I have time to do so.  There is a cmake veraible to disable shared libs if you don't want the dll
~~~
cmake -DENABLE_SHARED=off
~~~

**Comment by bleunova** [comment]
Hi @jstebbins any update on this would be appreciated. :)

**Comment by jstebbins** [comment]
Conflicts resolved.

There were also a few bugs in the generation of libSvtAv1Dec that had to be fixed in order to link the static lib version.

**Comment by 1480c1** [comment]
One thing I would like to mention is that cmake comes with a variable called `BUILD_SHARED_LIBS` that would select shared or static for the add_library functions that don't have a specific mode set. You could use that instead of creating a custom variable

**Comment by jstebbins** [comment]
BUILD_SHARED_LIBS does one or the other.  There is no nice way that I know of using BUILD_SHARED_LIBS to say, I may want one, or the other, or both.

**Comment by 1480c1** [comment]
What do you mean? I'm a bit confused.

**Comment by 1480c1** [comment]
Do you mean some of them as static libs and some as shared?

**Comment by jstebbins** [comment]
There are 2 add_library instances. One for shared and one for static libraries.  By default, both are built.

**Comment by 1480c1** [comment]
I see, so by default it will build both static and shared. Which will get installed/linked?

**Comment by 1480c1** [comment]
I'm trying to understand what is the design

**Comment by jstebbins** [comment]
Both get installed by default.  If shared libs are enabled, the SvtApp's are linked against the shared lib.  If shared is disabled, the SvtApp's are linked against the static libs.  External programs that use the libraries would choose which they link to depending on the needs of the app.

**Comment by 1480c1** [comment]
One thing you can do is leave everything as it is but use that variable instead of the custom variable because it will only make the libs shared if shared or static is not stated explicitly. So since you have shared and static stated explicitly, it would work perfectly fine in this case

**Comment by 1480c1** [comment]
at least from what I have seen

**Comment by 1480c1** [comment]
I'm going to test it in a sec

**Comment by jstebbins** [comment]
Assume I left one call to add_library unspecified so that BUILD_SHARED_LIBS could control the type of the library. What would I set in the other add_library call?  The only solution that makes any sense that I can think of would look like this:
~~~
if(BUILD_SHARED_LIBS)
    # Encoder Lib Source Files
    add_library(SvtAv1EncShared SHARED
        ${all_files}
        $<TARGET_OBJECTS:COMMON_CODEC>
        $<TARGET_OBJECTS:COMMON_C_DEFAULT>
        $<TARGET_OBJECTS:COMMON_ASM_SSE2>
        $<TARGET_OBJECTS:COMMON_ASM_SSSE3>
        $<TARGET_OBJECTS:COMMON_ASM_SSE4_1>
        $<TARGET_OBJECTS:COMMON_ASM_AVX2>
    )
    set_target_properties(SvtAv1EncShared PROPERTIES OUTPUT_NAME SvtAv1Enc)
    target_link_libraries(SvtAv1EncShared ${PLATFORM_LIBS})
    install(TARGETS SvtAv1EncShared DESTINATION "${CMAKE_INSTALL_LIBDIR}")
endif(BUILD_SHARED_LIBS)

add_library(SvtAv1EncStatic STATIC
    ${all_files}
    $<TARGET_OBJECTS:COMMON_CODEC>
    $<TARGET_OBJECTS:COMMON_C_DEFAULT>
    $<TARGET_OBJECTS:COMMON_ASM_SSE2>
    $<TARGET_OBJECTS:COMMON_ASM_SSSE3>
    $<TARGET_OBJECTS:COMMON_ASM_SSE4_1>
    $<TARGET_OBJECTS:COMMON_ASM_AVX2>
)
~~~
I could leave off the 'SHARED' in the add_library call above, but that's kind of pointless and obscures what the code is really doing.  So it would allow you to use the variable BUILD_SHARED_LIBS, but it doesn't use the built-in behaviour of add_library since it's not necessary and just obscures the meaning of the code.

Note that the only difference between this and what I've done in the PR is in the 'if' conditional.

**Comment by 1480c1** [comment]
That's what I mean

**Comment by 1480c1** [comment]
That way both static and shared can be built and assuming that the rest of the logic is fine, it should be as you written

**Comment by 1480c1** [comment]
The only real reason why I prefer to use `BUILD_SHARED_LIBS` is because that way I don't have to remember if a certain package used a special variable to do certain stuff.

**Comment by 1480c1** [comment]
Changing all of the `ENABLE_SHARED` variables to `BUILD_STATIC_LIBS` worked and both shared and static are built and with it enabled, the EncApp is linked with the shared objects

**Comment by jstebbins** [comment]
OK, I'll update the PR.  Thanks.

**Comment by 1480c1** [comment]
When running with it off, the app is statically built, although it still does link to shared pthread, m, c, vdso, and ld which might be a bit annoying

**Comment by 1480c1** [comment]
One thing that can be done is <https://stackoverflow.com/a/29824424> so that way it would be a little cleaner

**Comment by jstebbins** [comment]
If you are referring to object libraries, I'm already using them. Previously, it built static libs for the avx, sse, etc. dsp functions.  Then it linked those static libs into the shared lib.  You can't directly add a static lib to another static lib, so it was necessary to replace those with object libraries.

**Comment by 1480c1** [comment]
I was thinking more in the codec cmakelist using the `OBJECT IMPORTED` keyword for add_library, but trying it, it gave me undefined references, so nevermind

**Comment by jstebbins** [comment]
Conflicts fixed again...

File: Source/API/EbSvtAv1Time.h [hunk 1]
```diff
@@ -6,14 +6,24 @@
...
+
 #define NANOSECS_PER_SEC ((uint32_t)(1000000000L))
 
-void EbStartTime(uint64_t *Startseconds, uint64_t *Startuseconds);
-void EbFinishTime(uint64_t *Finishseconds, uint64_t *Finishuseconds);
-void EbComputeOverallElapsedTime(uint64_t Startseconds, uint64_t Startuseconds, uint64_t Finishseconds, uint64_t Finishuseconds, double *duration);
-void EbComputeOverallElapsedTimeMs(uint64_t Startseconds, uint64_t Startuseconds, uint64_t Finishseconds, uint64_t Finishuseconds, double *duration);
-void EbSleep(uint64_t milliSeconds);
-void EbInjector(uint64_t processedFrameCount, uint32_t injector_frame_rate);
+EB_API void EbStartTime(uint64_t *Startseconds, uint64_t *Startuseconds);
```
Comment (hassount) [inline Source/API/EbSvtAv1Time.h]: I understand the need to combine these calls into one, but I don't think that we will need to make it an API as such has other consequences. maybe we can keep them with different names at this point (one with a prefix eb then the app side would be without an eb prefix).

File: Source/Lib/Encoder/Codec/EbEncHandle.c [hunk 2]
```diff
@@ -275,8 +228,8 @@ EbErrorType InitThreadManagmentParams() {
 #elif defined(__linux__)
     const char* PROCESSORID = "processor";
     const char* PHYSICALID = "physical id";
-    int processor_id_len = strnlen_ss(PROCESSORID, 128);
-    int physical_id_len = strnlen_ss(PHYSICALID, 128);
+    int processor_id_len = strnlen(PROCESSORID, 128);
```
Comment (hassount) [inline Source/Lib/Encoder/Codec/EbEncHandle.c]: we can't use strnlen as it is an unsafe c function. We'll need to use strnlen_ss

File: Source/API/EbSvtAv1Time.h (hunk 1 above)
Comment (jstebbins) [reply Source/API/EbSvtAv1Time.h]: If there are no objections, I'll do as you suggest.

File: Source/Lib/Encoder/Codec/EbEncHandle.c (hunk 2 above)
Comment (jstebbins) [reply Source/Lib/Encoder/Codec/EbEncHandle.c]: Then I'll need to also duplicate these functions and give them a different name as you suggested for the time functions above.  I was told previously that you don't want these string functions to be part of the library API.

Comment (hassount) [reply Source/Lib/Encoder/Codec/EbEncHandle.c]: Sure, let's do so, eb inside the library, no eb at the application

File: Source/API/EbSvtAv1Time.h (hunk 1 above)
Comment (hassount) [reply Source/API/EbSvtAv1Time.h]: @andrey-norkin any issues with this ?

**Comment by jstebbins** [comment]
Last 2 commits should satisfy your requests...

**Comment by 1480c1** [comment]
Please make sure to check the ffmpeg patches, they have references to EbSvtAv1Time.h

**Comment by 1480c1** [comment]
Either modify those directly or ask @jsunintel to generate new patches with the right headers to add to this pr.

**Comment by jstebbins** [comment]
Yes, I see the failures.  Working on fixing...

**Comment by EwoutH** [comment]
@hassount @kylophone Ready to merge?

**Comment by EwoutH** [comment]
@kylophone Could you please review?

**Comment by hassount** [comment]
@jstebbins 
Would you be able to add a little switch to swap between building shared and or static libraries in build.sh ?

**Comment by jstebbins** [comment]
Sorry for the delay. Was travelling.

I have a build script I wrote that's based on Build/linux/build.sh, but I've modified it quite a lot.  Generalized it to make mingw cross compiling easier and such.  I can replace the current linux script with that and let you have a look to see if it's acceptable.

**Comment by jstebbins** [comment]
I'm confused by the travis failure that just appeared.  The build log indicates an error in code that does not appear in my local copy of this branch.  I.e. mrp_mode is not referenced in my copy of Source/Lib/Common/Codec/EbModeDecision.c

**Comment by jstebbins** [comment]
My last commit fixes an  error in master that is unmasked by my changes...

**Comment by hassount** [comment]
@jstebbins 
can you reproduce the travis build failure on your end ?

**Comment by jstebbins** [comment]
Yes.  I'll see if I can track it down.  I won't have time to look at it till Friday though.

**Comment by jstebbins** [comment]
I had a little time to investigate.  It looks to me like a bad organization of ```#define``` and header file dependencies.  MEMORY_FOOTPRINT_OPT_ME_MV is defined in Lib/Common/Codec/EbDefinitions.h.  API/EvSvtAv1Enc.h uses MEMORY_FOOTPRINT_OPT_ME_MV.  App includes API/EvSvtAv1Enc.h, but does not (and can not) include EbDefinitions.h.  So structure layout is different between what the app expects and what the library expects.

So it appears to be a problem in git head that is exposed by my build changes.

**Comment by hassount** [comment]
> I had a little time to investigate. It looks to me like a bad organization of `#define` and header file dependencies. MEMORY_FOOTPRINT_OPT_ME_MV is defined in Lib/Common/Codec/EbDefinitions.h. API/EvSvtAv1Enc.h uses MEMORY_FOOTPRINT_OPT_ME_MV. App includes API/EvSvtAv1Enc.h, but does not (and can not) include EbDefinitions.h. So structure layout is different between what the app expects and what the library expects.
> 
> So it appears to be a problem in git head that is exposed by my build changes.

Good catch, let's fix the master, this will hopefully address the merge conflicts.

**Comment by jstebbins** [comment]
> The command "wget -nc http://github.com/OpenVisualCloud/SVT-AV1-Resources/raw/master/video.tar.gz" failed and exited with 8 during

Build failure appears to be a network issue

**Comment by 1480c1** [comment]
I seems to be repo wide since https://github.com/OpenVisualCloud/SVT-AV1/pull/264 also failed due to similar issues

**Comment by jstebbins** [comment]
Resolved recent conflicts and updated build script with unit test options

**Comment by hassount** [comment]
@kylophone @joelsole 
Any issues with this ?

**Comment by hassount** [comment]
@jstebbins ready to merge ? 

@kylophone @joelsole any issues with this PR ?

**Comment by 1480c1** [comment]
Although this is a bit late, would it be possible to add some way to disable building shared entirely through something like a separate variable? It doesn't have to be in this pr

**Comment by jstebbins** [comment]
@hassount I believe so.  I discovered that last little bit while updating my mingw branch to fix building of unit tests.

@1480c1 ```build.sh -x``` disables building shared library.  Only static libSvtAv1Enc.a (and libSvtAv1Dec.a) is built and the binary is linked to it.  If you are talking about linking only static *system* libs, that's another PR I think with it's own variable to enable. It *should* simply be a matter of adding ```-static``` to ```CMAKE_EXE_LINKER_FLAGS```.  I did a quick test by manually adding this in the top CMakeLists.txt file and I got a completely static binary.

**Comment by 1480c1** [comment]
>  It should simply be a matter of adding -static to CMAKE_EXE_LINKER_FLAGS. I did a quick test by manually adding this in the top CMakeLists.txt file and I got a completely static binary.

Nice.

**Comment by kylophone** [comment]
Yeah, let's merge it. @hassount, I'll let you do the honours.

**Pull request merged**